python_modules/dagster/dagster_tests/asset_defs_tests/test_asset_deps.py

```diff
@@ -0,0 +1,420 @@
+    assert res.success
+
+
+def test_source_asset_deps_via_string():
+    a_source_asset = SourceAsset("a_key")  # noqa: F841
+
+    @asset(deps=["a_key"])
+    def depends_on_source_asset():
+        return None
+
+    assert len(depends_on_source_asset.input_names) == 1
+    assert depends_on_source_asset.op.ins["a_key"].dagster_type.kind == DagsterTypeKind.NOTHING
+
+    res = materialize([depends_on_source_asset], resources={"io_manager": TestingIOManager()})
+    assert res.success
+
+
+def test_source_asset_deps_via_key():
+    a_source_asset = SourceAsset("a_key")  # noqa: F841
+
+    @asset(deps=[AssetKey("a_key")])
+    def depends_on_source_asset():
+        return None
+
+    assert len(depends_on_source_asset.input_names) == 1
+    assert depends_on_source_asset.op.ins["a_key"].dagster_type.kind == DagsterTypeKind.NOTHING
+
+    res = materialize([depends_on_source_asset], resources={"io_manager": TestingIOManager()})
+    assert res.success
+
+
+def test_interop():
+    @asset
+    def no_value_asset():
+        return None
+
+    @asset(io_manager_key="fs_io_manager")
+    def value_asset() -> int:
+        return 1
+
+    @asset(
+        deps=[no_value_asset],
+    )
+    def interop_asset(value_asset: int):
+        assert value_asset == 1
+
+    assert len(interop_asset.input_names) == 2
+    assert interop_asset.op.ins["no_value_asset"].dagster_type.kind == DagsterTypeKind.NOTHING
+    assert interop_asset.op.ins["value_asset"].dagster_type.kind == DagsterTypeKind.SCALAR
+
+    res = materialize(
+        [no_value_asset, value_asset, interop_asset],
+        resources={"io_manager": TestingIOManager(), "fs_io_manager": FilesystemIOManager()},
+    )
+    assert res.success
+
+
+def test_non_existent_asset_key():
+    @asset(deps=["not_real"])
+    def my_asset():
+        return None
+
+    res = materialize([my_asset], resources={"io_manager": TestingIOManager()})
+
+    assert res.success
+
+
+def test_bad_types():
+    class NotAnAsset:
+        def __init__(self):
+            self.foo = "bar"
+
+    not_an_asset = NotAnAsset()
+
+    with pytest.raises(
+        DagsterInvalidDefinitionError,
+        match=(
+            "Cannot pass an instance of type <class"
+            " 'dagster_tests.asset_defs_tests.test_asset_deps.test_bad_types.<locals>.NotAnAsset'>"
+            " to deps parameter of @asset"
+        ),
+    ):
+
+        @asset(deps=[not_an_asset])
+        def my_asset():
+            return None
+
+
+def test_dep_via_deps_and_fn():
+    @asset
+    def the_upstream_asset():
+        return 1
+
+    with pytest.raises(
+        DagsterInvalidDefinitionError,
+        match=(
+            "@op 'depends_on_upstream_asset' decorated function has parameter 'the_upstream_asset'"
+            " that is one of the input_defs of type 'Nothing' which should not be included since no"
+            " data will be passed for it."
```

Comment:
ya we should definitely stack a PR to improve this I think 

maybe leave a comment here indicating this is known sub optimal

Reply:
opened an issue too https://github.com/dagster-io/dagster/issues/15167